highway.tex

```diff
@@ -24,807 +26,1323 @@
+To guarantee the above condition the simplest (but perhaps not the most efficient) way would be to instruct the honest validators to broadcast a set of $>n/2$ endorsements of a particular unit, after it becomes endorsed in their local view.
+%
+Various more efficient strategies here are possible, we discuss some more practical approaches to endorsements in Section~\ref{sec:practical}.
 
-\begin{proof}[Proof of Theorem~\ref{thm:limit_spam}]
-During the timespan of a single era, each chain can grow to the height of at most $2^{K-k_{min}}$, as no validator is allowed to create messages more often than the minimum round exponent.
 
-Due to the Lemma \ref{chainBoundLemma}, in $J(\mu)$ there can be at most $n$ chains of each of at most $d$ dishonest validators, and additionally one chain of each of the remaining validators (as all messages of an honest validator are bound to form a chain), so $(d\cdot n + n-d)$ chains in total.
+\section{Practical considerations}\label{sec:practical}
+\subsection{Dynamic Round Lengths}
+When defining the base version of our protocol we divided the time into fixed length rounds.
+%
+The length of a round was picked so as to make sure that in each such round, even assuming worst case delays, a few communication "round trips" between honest validators would be possible.
+%
+On the theoretical side, the worst case message delay $\Delta$ was available from the partial synchrony assumption in the {\it known $\Delta$ flavour}, as initially defined in~\cite{DLS88}.
+%
+In practice, one can estimate an appropriate $\Delta$ using experiments, yet there are a few dangers to be aware of when setting a constant $\Delta$:
+\begin{itemize}
+    \item If we set the $\Delta$ too optimistic (i.e., too low) then liveness might be violated, as some messages between (high ping) pairs of validators ping can arrive too late.
+    \item If we set the $\Delta$ too pessimistic, then the protocol might become quite slow, as during significant portion of the round, the validators would stay idle.
+\end{itemize}
 
-In summary, at most $2^{K-k_{min}}\cdot (d\cdot n + n-d)$ messages can be in $J(\mu)$. 
+This motivates the use of dynamic round lengths so that the protocol by itself adapts to the current network conditions.
+%
+At a theoretical layer this "approach" corresponds to the {\it unknown $\Delta$ flavour} in the original definition of partial synchrony by~\cite{DLS88}.
+%
+The theoretical model states that there is some fixed, worst case message delay $\Delta$, yet it is not available at the start of the protocol execution.
+%
+Even though these two models of partial synchrony are equivalent via almost black-box reductions, we present here a custom version of our protocol for the unknown $\Delta$ model that is optimized for practical implementations.
 
-\end{proof}
+\noindent {\bf Rules for changing round lengths.}
+We measure time in milliseconds since the epoch\footnote{Unix time: number of milliseconds since the beginning of 1970, UTC}.
```

Comment:
Normally "Unix time" is _seconds_ since the epoch… maybe something like:
```suggestion
We measure time in milliseconds since the epoch\footnote{Unix time: number of seconds since the beginning of 1970, UTC, but millisecond resolution is more suitable for this purpose}.
```